autorest/azure/devicetoken_test.go

```diff
@@ -0,0 +1,222 @@
+	client := &autorest.Client{Sender: sender}
+
+	_, err := InitiateDeviceAuth(client, TestClientID, TestResource)
+	if err == nil {
+		t.Errorf("failed to get error") // (expecting Failed to decode Device Authorization Code json)
+	}
+}
+
+func deviceCode() *DeviceCode {
+	var deviceCode DeviceCode
+	json.Unmarshal([]byte(MockDeviceCodeResponse), &deviceCode)
+	deviceCode.Resource = to.StringPtr("resource")
+	return &deviceCode
+}
+
+func TestDeviceTokenReturns(t *testing.T) {
+	sender := mocks.NewSender()
+	sender.EmitContent(MockDeviceTokenResponse)
+	sender.EmitStatus("OK", 200)
+	client := &autorest.Client{Sender: sender}
+
+	_, err := WaitForUserCompletion(client, TestClientID, deviceCode())
+	if err != nil {
+		t.Errorf("got error unexpectedly")
+	}
+}
+
+func TestDeviceTokenReturnsErrorIfSendingFails(t *testing.T) {
+	sender := mocks.NewSender()
+	sender.EmitErrors(1)
+	sender.SetError(fmt.Errorf("this is an error"))
+	client := &autorest.Client{Sender: sender}
+
+	_, err := WaitForUserCompletion(client, TestClientID, deviceCode())
+	if err == nil {
+		t.Errorf("failed to get err") // (expecting Error occurred while exchanging device code for token)
+	}
+}
+
+func TestDeviceTokenReturnsErrorIfServerError(t *testing.T) {
+	sender := mocks.NewSender()
+	sender.EmitStatus("Internal Server Error", 500)
+	client := &autorest.Client{Sender: sender}
+
+	_, err := WaitForUserCompletion(client, TestClientID, deviceCode())
+	if err == nil {
+		t.Errorf("failed to get error") // (expecting Exchanging Device Code for OAuth token resulted in Status)
+	}
+}
+
+func TestDeviceTokenReturnsErrorIfCannotDeserializeDeviceToken(t *testing.T) {
+	gibberishJSON := strings.Replace(MockDeviceTokenResponse, "expires_in", ";:\"gibberish", -1)
+	sender := mocks.NewSender()
+	sender.EmitContent(gibberishJSON)
+	client := &autorest.Client{Sender: sender}
+
+	_, err := WaitForUserCompletion(client, TestClientID, deviceCode())
+	if err == nil {
+		t.Errorf("failed to get err") // (expecting Error occurred while exchanging device code for token)
+	}
+}
+
+func errorDeviceTokenResponse(message string) string {
+	return `{ "error": "` + message + `" }`
+}
+
+func TestDeviceTokenReturnsErrorIfAuthorizationPending(t *testing.T) {
+	sender := mocks.NewSender()
+	sender.EmitContent(errorDeviceTokenResponse("authorization_pending"))
+	sender.EmitStatus("Bad Request", 400)
+	client := &autorest.Client{Sender: sender}
+
+	go func() {
+		time.Sleep(1 * time.Second)
+		sender.EmitContent(errorDeviceTokenResponse("access_denied"))
+	}()
+
+	_, _ = WaitForUserCompletion(client, TestClientID, deviceCode())
+}
+
+func TestDeviceTokenReturnsErrorIfSlowDown(t *testing.T) {
+	sender := mocks.NewSender()
+	sender.EmitContent(errorDeviceTokenResponse("slow_down"))
+	sender.EmitStatus("Bad Request", 400)
+	client := &autorest.Client{Sender: sender}
+
+	go func() {
+		time.Sleep(1 * time.Second)
+		sender.EmitContent(errorDeviceTokenResponse("access_denied"))
+	}()
+
```

Comment:
(ditto as well) "ignoring this in favor of Ahmet's solution and removing the sleep entirely."